File: android/core/src/main/kotlin/com/aerodue/core/flight/Matcher.kt
package com.aerodue.core.flight

import com.aerodue.core.domain.FlightLeg
import com.aerodue.core.domain.FlightStatus
import com.aerodue.core.domain.Trip
import com.aerodue.core.location.GeoFix
import com.aerodue.core.location.nearestAirportIcao
import java.time.Duration

fun matchTripFromLocation(
    fix: GeoFix,
    candidateLegs: List<FlightLeg>,
    windowHours: Long = 18,
): Trip? {
    val airport = nearestAirportIcao(fix) ?: return null
    val now = fix.recordedAt
    val window = Duration.ofHours(windowHours)

    val matches = candidateLegs.filter { leg ->
        val depNear = leg.departureAirport == airport
        val arrNear = leg.arrivalAirport == airport
        val inWindow =
            !now.isBefore(leg.scheduledDeparture.minus(window)) &&
                !now.isAfter(leg.scheduledArrival.plus(window))
        inWindow && (depNear || arrNear)
    }
    if (matches.isEmpty()) return null

    val sorted = matches.sortedBy { leg ->
        kotlin.math.abs(Duration.between(leg.scheduledDeparture, now).seconds)
    }
    val first = sorted.first()
    return Trip(tripId = "auto-${first.flightNumber}", legs = sorted)
}

fun classifyDelayMinutes(leg: FlightLeg): Int? {
    val actual = leg.actualDeparture ?: return null
    if (leg.status !in setOf(
            FlightStatus.DELAYED,
            FlightStatus.DEPARTED,
            FlightStatus.ARRIVED,
        )
    ) {
        return null
    }
    val delta = Duration.between(leg.scheduledDeparture, actual)
    return maxOf(0, delta.toMinutes().toInt())
}

File: backend/aerodue/core/flight/matcher.py
from datetime import datetime, timedelta

from aerodue.core.location.gps import GeoFix, nearest_airport_icao
from aerodue.domain.flight import FlightLeg, FlightStatus, Trip


def match_trip_from_location(
    fix: GeoFix,
    candidate_legs: list[FlightLeg],
    window_hours: int = 18,
) -> Trip | None:
    """
    Pick the most likely active leg using airport proximity and schedule window.
    """
    airport = nearest_airport_icao(fix)
    if not airport:
        return None

    now = fix.recorded_at
    window = timedelta(hours=window_hours)
    matches: list[FlightLeg] = []

    for leg in candidate_legs:
        dep_near = leg.departure_airport == airport
        arr_near = leg.arrival_airport == airport
        in_window = (
            leg.scheduled_departure - window <= now <= leg.scheduled_arrival + window
        )
        if in_window and (dep_near or arr_near):
            matches.append(leg)

    if not matches:
        return None

    matches.sort(key=lambda l: abs((l.scheduled_departure - now).total_seconds()))
    return Trip(trip_id=f"auto-{matches[0].flight_number}", legs=matches)


def classify_delay_minutes(leg: FlightLeg) -> int | None:
    if leg.actual_departure is None or leg.status not in (
        FlightStatus.DELAYED,
        FlightStatus.DEPARTED,
        FlightStatus.ARRIVED,
    ):
        return None
    delta = leg.actual_departure - leg.scheduled_departure
    return max(0, int(delta.total_seconds() // 60))
